### PR TITLE
Inline function arguments correctly

### PR DESCRIPTION
**Pull request by SaarYogev**
Currently it replaces function parameters with the function arguments.  Still need to make sure arguments will be resolved in the same order were they not inlined by creating variables for them before the main inline. Here's a Kotlin example:
That's the original file. We'll inline `foo()`
```kotlin
var count = 1
private fun main() {
    val a = foo(bar())
}

private fun foo(number: Int): Int {
    return number * bar() + number
}

private fun bar(): Int {
    return count++
}
```
The original code should get `2 * 3 + 2` inside foo.
My code currently inlines the values directly, like so:
```kotlin
var count = 1
private fun main() {
    val a = bar() * bar() + bar()
}

    private fun bar(): Int {
    return count++
}
```
Which leads the calculation to be `2 * 3 + 4`, because the function will be called an additional time.
The correct inline (as shown by Kotlin) is:
```kotlin
var count = 1
private fun main() {
    val number = bar()
    val a = number * bar() + number
}

    private fun bar(): Int {
    return count++
}
```
Which makes sure number has a single value, assined before any calls for `bar()` happening inside `foo()`.

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/refactoring/inline/RsInlineFunctionProcessor.kt [hunk 1]
```diff
@@ -222,6 +224,22 @@ class RsInlineFunctionProcessor(
         }
     }
 
+    private fun replaceParametersWithArgumentValues(functionDup: RsFunction, caller: PsiElement, funcScope: LocalSearchScope) {
+        val parameterBindingsList = functionDup.valueParameters
+            .filter { it !is RsSelfParameter }
+            .map { it.pat }
+            .filterIsInstance(RsPatIdent::class.java)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/refactoring/inline/RsInlineFunctionProcessor.kt:231]:
```suggestion
            .filterIsInstance<RsPatIdent>()
```

**Review by vlad20012 — COMMENTED** [review]
This wouldn't work with complex patterns in parameters arguments, e.g.
```rust
fn foo((a, b): (i32, i32), c: i32) {}
```
Let's `zip` `functionDup.valueParameters` with `valueArgumentList` directly, and then (for now) just skip parameters with such complex patterns. And let's add a test for this.

Also, I'm not sure how this will work for methods (with `self`!) invoked with UFCS syntax, e.g.
```rust
impl S {
    fn foo(&self, b: i32) {}
}
...
S::foo(S, 42);
```

Let's also add a test for this.

File: src/main/kotlin/org/rust/ide/refactoring/inline/RsInlineFunctionProcessor.kt [hunk 2]
```diff
@@ -276,6 +294,18 @@ class RsInlineFunctionProcessor(
         }
     }
 
+    private fun replaceParamWithExpr(param: RsPatBinding, argument: RsExpr, funcScope: LocalSearchScope) {
+        val argumentText = buildString {
+            append(argument.text.removePrefix("mut ").removePrefix("&"))
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/refactoring/inline/RsInlineFunctionProcessor.kt:299]: Let's add a test where these `removePrefix` are needed

**Comment by Kobzol** [comment]
@SaarYogev If you're going to handle arguments with side expressions in this PR (i.e. creating variables with `let ...` before the function call), then I think you can also handle complex patterns right away, because the solution for them is the same:
```rust
fn bar() {
    foo((1, 1));
    // turns into
    let (a, b) = (1, 1);
    println!("{}", a);
}
fn foo((a, b): (i32, i32)) {
    println!("{}", a);
}
```
For both cases when you introduce the `let` binding you should check if it doesn't conflict with existing variables in the function where the inlining was invoked (and use a different name in that case?).

**Comment by SaarYogev** [comment]
Now I am creating let declarations for arguments with no constant value (not primitives or variables alone).
the only faling test is failing because of a mutability check. This function inside `test inline function with differently named reference and mutable function parameters`

```rust
fn foo(i: &i32, j: &mut i32) -> i32 {
    return i + j + bar();
}
```

has j, which is a mutable reference. However, `parameterBinding.mutability.isMut` returns `false` for him. It seems like a bug, as I wasn't able to find another place where it shown as mutable.

**Comment by Kobzol** [comment]
It doesn't seem like a bug, j is not a mutable binding. It is an immutable binding of type "mutable referenxe to i32".

**Comment by SaarYogev** [comment]
So I guess I need to get the actual mutability from the type data?

**Comment by Kobzol** [comment]
Hmm, why do you even need any special handling of `mut` bindings? If the parameter was `mut`, the created let declaration should be `mut` (and vice versa). Is there any problem with this?

**Comment by SaarYogev** [comment]
@Kobzol No, I just don't know how to get to the parameter instead of the binding.

**Comment by Kobzol** [comment]
Sorry, I don't understand what do you mean by that. Could you show me an example?

**Comment by SaarYogev** [comment]
@Kobzol You told me that I should check if the parameters is *mutable* or not (which I assumed will be easy, something like `function.parameters[0].isMut`), but I only found how to see the properties of the parameterBindings, which as you said won't help me figure it out (using `function.valueParameters`). I would like to know how can I check the mutibility of function parameters, given an `RsFcuntion`.

**Comment by Kobzol** [comment]
Ok, so we had a slight misunderstanding I think. In the test that you have mentioned, the `j` binding should NOT be `mut`, because the corresponding parameter is also not `mut`. In other words, I think that checking only `parameterBinding.mutability.isMut` is enough. If it returns true, create the local binding with `mut`. If not, create an immutable binding.

```rust
fn main() {
    let i = bar();
    let j = qux(); // this should NOT be a mutable binding
    let a = i + j + bar();
}
fn bar() -> &i32 {
    let baz = 5;
    return &baz;
}
fn qux() -> &mut i32 {
    let mut quux = 5;
    return &mut quux;
}
```
However, this test does not compile (with rustc), please try to change it so that it does.

**Comment by SaarYogev** [comment]
@Kobzol I don't understand why, which parameter do you refer to in 

> the corresponding parameter

I see that `qux()` does return a mutable reference.
`j` doesn't have to be mutable to be added to `a`, but I assume we don't intend to scan extra code in the scope where we inline the function into, unless for very basic purposes (like avoiding duplicate variable names).
So I can't tell how would we know that this code isn't possible:
```rust
fn main() {
    let i = bar();
    let mut j = qux(); // making j mut
    j = j + 1; // using j's mutability to increment its value
    let a = i + j + bar();
}
fn bar() -> &i32 {
    let baz = 5;
    return &baz;
}
fn qux() -> &mut i32 {
    let mut quux = 5;
    return &mut quux;
}
```

**Comment by Kobzol** [comment]
`j = ...` is only possible if `j` was declared with `mut` in the original function that is being inlined. We just assume that the original function was correct. If it was correct, it should remain correct after inlining. If it was incorrect, we shouldn't "fix" the code when inlining it.

To sum up: I think that it is valid to create the new binding with `mut` ONLY if the original parameter binding had `mut`. In fact, the binding of the local variable should be exactly the same as the binding of the parameter.

Maybe there is some case where this wouldn't work, but I think that as a first approach it is a valid and simple heuristic.

**Comment by SaarYogev** [comment]
I agree that we should leave the function as is and not try to fix it, but the original function from the failing test has a mutable variable as far as I can tell: 
We inline `foo` like so: 
```rust
fn /*caret*/foo(i: &i32, j: &mut i32) -> i32 {
    return i + j + bar();
}
```

`j` is declared as `&mut`, so I do not understand what's the difference between the type and the binding that makes it appear immutable. I assumed that `j: &mut i32` should convert to `let mut j`, though I can see that reference is gone, don't know what to do about that.

**Comment by Kobzol** [comment]
The mutability of the binding and the type are two separate things. For example:
- `i: &u32` is an immutable binding with type "immutable reference to `u32`"
- `i: &mut u32` is an immutable binding with type "mutable reference to `u32`"
- `mut i: &u32` is a mutable binding with type "immutable reference to `u32`"
- `mut i: &mut u32` is a mutable binding with type "mutable reference to `u32`"

Therefore, `j: &mut i32` is an immutable binding, therefore it should be converted to `let j` and not `let mut j`.

**Comment by SaarYogev** [comment]
I think I do understand now, it's like changing the contents of the object (mutable object) versus changing the reference to a different object (mutable object reference) in an object-oriented language. I'll try to test that now, thanks 😄

**Comment by Kobzol** [comment]
Exactly, that is a good comparison! :)

**Comment by SaarYogev** [comment]
@Kobzol All done, was a pretty easy fix. Now the test is checking both immutable binding of a mutable var, or a mutable binding of an immutable var.

File: src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt [hunk 3]
```diff
@@ -866,11 +867,17 @@ class RsInlineFunctionTest : RsTestBase() {
                 let mut quux = 5;
                 return &mut quux;
             }
+
+            fn corge() -> &mut i32 {
```
Comment (Kobzol) [inline src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt:871]: Could you also please make the test code compilable with rustc? It could be a bit hard with `&mut` references (although you could use &'static refs and some unsafe code), maybe choose some simpler code snippet?

Comment (SaarYogev) [reply src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt:871]: I can make `corge()` return a plain `i32` that will have a mutable binding later and be done with that part, but I do need to handle the errors in `qux()`, where a mutable reference is the point. 
I'll try to handle the lifetimes as the compiler suggests, but it might take me some time.

Comment (SaarYogev) [reply src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt:871]: Well, I have a piece of test code that passes rustc. However, it doesn't pass our own checks, saying `Cannot move` under `*i` and `*j`
```rust
fn main() {
    let _a = foo(&bar(), &mut qux(), corge());
}

#[allow(unused_mut)]
fn /*caret*/foo(i: &i32, j: &mut i32, mut k: i32) -> i32 {
    return *i + *j + k + bar();
}

fn bar() -> i32 {
    let baz = 5;
    return baz;
}

fn qux() -> i32 {
    let quux = 5;
    return quux;
}

fn corge() -> i32 {
    let grault = 5;
    return grault;
}
```

Comment (Kobzol) [reply src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt:871]: Try to add `@ProjectDescriptor(WithStdlibRustProjectDescriptor::class)` above the test function. I think that it needs stdlib to see macros that implement `Copy` for primitive types.

Comment (SaarYogev) [reply src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt:871]: It doesn't help unfortunately, both because I'm using *Inject Language* so it already knows it's *Rust*, and because I tried it all in a scratch file first, so there it's pure *Rust*. I guess that it's a bug in the plugin.

Comment (Kobzol) [reply src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt:871]: I checked out your branch and the test passes, so it's OK. This code also works in the plugin if you use it inside some project.

Comment (SaarYogev) [reply src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt]: Did you use in the tests what I commited or the code I supplied above? because I only committed it now, the test doesn't pass on my machine, and if it didn't work in a scratch file it shouldn't work in a project either (unless cargo related).

Comment (Kobzol) [reply src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt]: Sorry, didn't see the last commit. But still, the code works inside a project. It needs access to a project and to stdlib to infer `Copy` properly. Inside injected Rust strings and scratch files it does not have access to a project, that's why it shows the `Cannot move` false positive.

image

Comment (SaarYogev) [reply src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt]: That's good, means the code itself, doesn't need change any more.
I also see from searching that `@ProjectDescriptor(WithStdlibRustProjectDescriptor::class)` is the standard way to introduce `stdlib` inside, any idea why it's not working?

Comment (Kobzol) [reply src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt]: If the problem that you're trying the solve is that there are red annotations inside the injected Rust test code, then the project descriptor will not solve it. But it's not a problem by itself, such annotations are present in a lot of tests, it's completely fine. Do you have any other problems with this test?

Comment (SaarYogev) [reply src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt]: No, I just assumed there is a way to solve it 😅

**Comment by SaarYogev** [comment]
@Kobzol All fixed, now the test both passes `rustc` and the test itself, and the change also revealed a gap in my type check regarding unary expressions, which now I strip of their unary operator, as should be, and then checking the inner expression.

**Comment by SaarYogev** [comment]
@Kobzol Anything left for me to do here? If not I'll start working on #6003 #6005 in the meantime😄.

**Comment by Kobzol** [comment]
Could you add tests for some more complex patterns? For example
```
fn foo((a, b): (u32, u32)) { ... }
```

**Comment by SaarYogev** [comment]
I wish I could do that, each implementation of `RsPat` might require a different way of getting variable name, mutibility, etc.
I thought of handling this using a strategy for each special `RsPat` case (and one for the default case), but it seems overkill and not going well with the general code style in the repo.

**Comment by Kobzol** [comment]
Hmm, it doesn't seem that bad to me to create a `when` statement for all of them. Actually if we won't check existing variable names in this PR, you can probably just copy the `text` of the parameter binding?

@vlad20012 WDYT?

**Comment by SaarYogev** [comment]
@Kobzol There's still the issue of literal/non-literal values and, and I made a different function for each. The reason is that literal values (e.g. `foo(6);`, or `let a = 6; foo(a);`) don't need to be extracted to a *let declaration*, and it looks silly if they do (and makes our inline less concise):
image

**Comment by Kobzol** [comment]
I suppose that we could check if the binding of the local variable is exactly the same as the binding of the parameter (they can differ in slight details, for example mutability) and remove the new `let` declaration in such a case.

But I don't see this as a big problem, first the refactoring should support all possible cases, and then we can start focusing on generating prettier code.

**Comment by SaarYogev** [comment]
It would also happen in the first case that I described:
```rust
fn main() {
    foo(6);
}

fn /*caret*/foo(bar: i32) {
    println!("{}", bar);
}
```
our expected would be `println~("{}", 6);`, but actual will be `let bar = 6; println~("{}", bar);`, which isn't needed as `6` can't change value like a function call.

As for not choosing names as best as possible to make it more generic, it's not exactly cosmetic, as I use this to split into 2 variables in the tuple deconstruction. However, it is using `parameter.text`, which is what you suggested for making it generic. The issue is that I don't know if that would work for every single type.

File: src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt [hunk 4]
```diff
@@ -968,6 +968,60 @@ class RsInlineFunctionTest : RsTestBase() {
...
+            }
+
+            fn bar() -> S {
+                S{ a:5, b:5 }
+            }
+
+            struct S { a: i32, b: i32 }
+            """, """
+            fn main() {
+                let _ = bar();
```
Comment (Kobzol) [inline src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt:1012]: IMO this should create
```rust
 fn main() {
      let S { a, b } = bar();
      let _c = a + b + bar().a;
}
```

Comment (SaarYogev) [reply src/test/kotlin/org/rust/ide/refactoring/RsInlineFunctionTest.kt:1012]: agree, fixed.

**Comment by SaarYogev** [comment]
@Kobzol Fixed most things, taking the "variable name" in the parameter as the variable name in the new *let* declaration works surprisingly well 😄 . 
I do have one problematic thing left, related to not adding unnecessary `let` declarations - 
`test inline function with differently named function parameters with inlined references`
Eexpected:
```rust
fn main() {
    let b = 6;
    let a = b + bar(&b);
}
```
Actual:
```rust
fn main() {
    let b = 6;
    let i = b;
    let a = i + bar(&i);
}
```
It adds a new let declaration (`let i`) because he doesn't if `6` is a unambiguous expression, or one that can change with each call.
The previous solution was this if statement:
`if (argument is RsPathExpr || argument is RsLitExpr || argument is RsUnaryExpr || argument is RsPatConst)`
But it wasn't generic, so I removed it as agreed.
Is there a generic way (either property/mehtod based or type-based) to know if an expression is unambiguous (won't change if referenced to twice, by reference or by just copying)?

File: src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt [hunk 5]
```diff
@@ -209,7 +217,50 @@ class RsInlineFunctionProcessor(
...
+            .filterIsInstance(RsPat::class.java)
+
+        val valueArgumentList = when (caller) {
+            is RsCallExpr -> caller.valueArgumentList
+            is RsDotExpr -> caller.methodCall?.valueArgumentList ?: return
+            else -> return
+        }
+        for ((index, parameter) in parameterBindingsList.withIndex()) {
+            var argument = valueArgumentList.exprList[index]
+            if (argument is RsUnaryExpr) {
```
Comment (Kobzol) [inline src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:233]: Why is this here? The following case produced wrong output:
```rust
fn foo(x: bool) {
    println!("{}", x);
}
fn main() {
    foo/*caret*/(!true);
}
```

File: src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt [hunk 6]
```diff
@@ -209,7 +217,50 @@ class RsInlineFunctionProcessor(
...
+                argument = argument.expr
+            }
+            replaceParameterWithLetDeclaration(parameter, argument, caller)
+        }
+    }
+
+    private fun replaceParameterWithLetDeclaration(parameter: RsPat, argument: RsExpr, caller: PsiElement) {
+        val name: String = parameter.text.removePrefix("mut ")
+        if (name != argument.text) {
```
Comment (Kobzol) [inline src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:242]: What does this condition do? Seems that it's not required. Argument probably won't be equal to e.g. `a: i32`.

File: src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt [hunk 7]
```diff
@@ -263,6 +314,26 @@ class RsInlineFunctionProcessor(
         }
     }
 
+    private fun replaceParamWithExpr(param: RsPat, argument: RsExpr, funcScope: LocalSearchScope) {
```
Comment (Kobzol) [inline src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:317]: It seems that this functions is unused?

File: src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt [hunk 8]
```diff
@@ -209,7 +217,50 @@ class RsInlineFunctionProcessor(
...
+            .map { it.pat }
+            .filterIsInstance(RsPat::class.java)
+
+        val valueArgumentList = when (caller) {
+            is RsCallExpr -> caller.valueArgumentList
+            is RsDotExpr -> caller.methodCall?.valueArgumentList ?: return
+            else -> return
+        }
+        for ((index, parameter) in parameterBindingsList.withIndex()) {
+            var argument = valueArgumentList.exprList[index]
```
Comment (Kobzol) [inline src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:232]: If there is a function call without enough arguments (e.g. `fn foo(a: u32)` and you call with `foo()`), the refactoring shows a warning (conflict dialog). The user can proceed with the warning however and in that case this code would crash. Assume that there may not be enough arguments and do not use `[]` indexing here.

**Review by Kobzol — CHANGES_REQUESTED** [review]
I would keep that for future PR, let's just always create the `let` declaration in this PR.

**Comment by Kobzol** [comment]
Btw `self` replacement does not seem to work for some cases:
```rust
struct S { a: u32, b: u32 }

impl S {
    fn foo(&self, a: u32) {
        println!("{} {}", self.a, a);
    }
}

fn main() {
    let s = S { a: 0, b: 0 };
    S::foo(&s, 1);
}
```
Maybe we could remove it and use your code for replacing both normal argument and `&self`?

File: src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt (hunk 6 above)
Comment (SaarYogev) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:242]: It is equal in about half the tests, that *if* statement prevents us from turning this
```rust
fn main() {
    let bar = 10i32;
    foo(bar);
}
```
into this
```rust
fn main() {
    let bar = 10i32;
    let bar = bar;
    println!("{}", bar);
}
```
and creating a *let* declaration where the exact same one exists.

Comment (Kobzol) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:242]: Sorry, I didn't notice that it compares only to the name of the parameter. Let's keep it then and see if it works in the wild.

File: src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt (hunk 7 above)
Comment (SaarYogev) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:317]: It isn't used, since we decided to always make a *let* declaration, and will be used again if we won't create it when unecessary besides the obvious `name != argument.text` ([explained here](https://github.com/intellij-rust/intellij-rust/pull/6004#discussion_r512216550)) that's already covered. I can remove it for now too if you want.

Comment (Kobzol) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:317]: Let's remove it for this PR then :)

File: src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt (hunk 8 above)
Comment (SaarYogev) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:232]: Didn't see a safe way to return (`[]`, `get()`, and `elementAt()` all throw exceptions), so I created an *if* statement that returns if `index` surpasses the list's size.

Comment (Kobzol) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:232]: You can use [this](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/get-or-null.html) :)

Comment (SaarYogev) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:235]: Yeah, I know about this and `getOrElse()`, they're just delaying the problem a single line, to then *return* when we see a *null*.
I prefer to *return* just as I see that  there aren't enough arguments.

File: src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt (hunk 7 above)
Comment (SaarYogev) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt]: Done

File: src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt (hunk 8 above)
Comment (Kobzol) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:235]: You could use something like `var argument = valueArgumentList.exprList.getOrNull(index) ?: return` or just return before the cycle, since the sizes are known beforehand, but that's a detail :)

File: src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt (hunk 5 above)
Comment (SaarYogev) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt:236]: It's supposed to remove `mut` and `&` (mut is handled seperately if needed, & is automatically dereferenced). I'll fix the if statement accordigly. It helps about 10 tests reach the state where `name == argument.text` and there's not *let* decleration creation.

Comment (SaarYogev) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt]: Fixed, now it it will only shorten the argument if all leafs are `&` or `mut`.

Comment (Kobzol) [reply src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt]: It still breaks basics scenarios:
```rust
fn foo(x: &i32) {
    println!("{}", *x);
}
fn main() {
    let a = 5;
    foo/*caret*/(&a);
}
```

You can't just arbitrarily remove unary expressions, therefore I suggest to remove this code entirely :)

**Comment by SaarYogev** [comment]
@Kobzol Checked it now, it happens when trying to use `::` syntax with a method (with `self` parameter). It doesn't realise the first argument is the same as `self` supplied from a `foo.bar()` syntax. Working on  a fix now.

**Comment by SaarYogev** [comment]
@Kobzol Fixed every requested change you had + your comment with the `::`, all is now tested and passes.
The last thing I pushed was handling *let declarations* with only *ref*, like you said [here](https://github.com/intellij-rust/intellij-rust/pull/6004#discussion_r513264458). 
For that I created code that renames the created *let declaration*,as we can't have stuff like `let vec = &vec;`, using the same name twice. The fixing code is quite messy currently, as renaming had to occur in the end, forcing me to inline all my previous changes this PR into a single function. I'll work on refactoring that back on somehing sensible now, while you can still see if the logic is good to merge.

**Comment by vlad20012** [comment]
Cool, it works! I'd refactored the code a bit, but it works fine and especially much better than before, so let's just merge it as is. Thanks!

@SaarYogev could you, please, squash these commits?

**Comment by vlad20012** [comment]
:+1: 
bors r+

**Comment by vlad20012** [comment]
bors retry
